src/aws-cpp-sdk-core/source/client/RequestCompression.cpp

```diff
@@ -0,0 +1,284 @@
+    if (algorithm == CompressionAlgorithm::GZIP)
+    {
+        // calculating stream size
+        input->seekg(0, input->end);
+        size_t streamSize = input->tellg();
+        input->seekg(0, input->beg);
+
+        AWS_LOGSTREAM_TRACE(AWS_REQUEST_COMPRESSION_LOG_TAG, "Uncompressing request of " << streamSize << " bytes.");
+
+        // Preparing output
+        std::shared_ptr<Aws::IOStream> output = Aws::MakeShared<Aws::StringStream>( AWS_REQUEST_COMPRESSION_ALLOCATION_TAG);
+        if(!output)
+        {
+            AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Failed to allocate output while uncompressing")
+            return false;
+        }
+
+        // Prepare ZLIB to uncompress
+        int ret = Z_NULL;
+        z_stream strm = {};
+        auto in = Aws::MakeUniqueArray<unsigned char>(ZLIB_CHUNK, AWS_REQUEST_COMPRESSION_ALLOCATION_TAG);
+        if(!in)
+        {
+            AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Failed to allocate in buffer while uncompressing")
+            return false;
+        }
+
+        auto out = Aws::MakeUniqueArray<unsigned char>(ZLIB_CHUNK, AWS_REQUEST_COMPRESSION_ALLOCATION_TAG);
+        if(!out)
+        {
+            AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Failed to allocate out buffer while uncompressing")
+            return false;
+        }
+        // Cleaning Zlib stream state flags
+        strm.zalloc = Z_NULL;
+        strm.zfree = Z_NULL;
+        strm.opaque = Z_NULL;
+        strm.avail_in = 0;
+        strm.next_in = Z_NULL;
+
+        const int MAX_WINDOW_GZIP = 31;
+        ret = inflateInit2(&strm, MAX_WINDOW_GZIP);
+        if (ret != Z_OK)
+        {
+            return false;
+        }
+
+        //Adding one to the stream size counter to account for the EOF marker.
+        streamSize++;
+        size_t toRead;
+        // Decompress
+        do {
+            toRead = (streamSize < ZLIB_CHUNK)?streamSize:ZLIB_CHUNK;
+            if (toRead < 1) break; // Nothing left to read
+            // Fill the buffer
+            if(! input->read(reinterpret_cast<char *>(in.get()), toRead))
+            {
+                if (input->eof())
+                {
+                    //skip passing the EOF to the buffer
+                    toRead--;
+                }
+                else
+                {
+                    AWS_LOGSTREAM_ERROR(
+                        AWS_REQUEST_COMPRESSION_LOG_TAG,
+                        "Compress request failed to read from stream");
+                    return false;
+                }
+            }
+
+            // Filling input buffer to decompress
+            strm.avail_in = toRead;
+            strm.next_in = in.get();
+            do
+            {
+                // Run inflate on buffers
+                strm.avail_out = ZLIB_CHUNK;
+                strm.next_out = out.get();
+
+                ret = inflate(&strm, Z_NO_FLUSH);
+                // Catch errors
```

Comment:
>>what if ret does not fall into any of the switch statement?

>Then it is not an error.

maybe worth catching all other unhanded errors

```
if (ret < 0)
{
    switch (ret)
    {
    case Z_NEED_DICT:
        AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Compressed request failed to inflate with code: Z_NEED_DICT");
        return false;
    case Z_DATA_ERROR:
        AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Compressed request failed to inflate with code: Z_DATA_ERROR");
        return false;
    case Z_MEM_ERROR:
        (void)inflateEnd(&strm);
        AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Compressed request failed to inflate with code: Z_MEM_ERROR");
        return false;
    case default:
        (void)inflateEnd(&strm);
        AWS_LOGSTREAM_ERROR(AWS_REQUEST_COMPRESSION_LOG_TAG, "Compressed request failed to inflate with unknown error code: " << ret);
        return false;
    }
}
```

Reply:
Tested it, it will catch also some internal non-error status if only checking it is not Z_OK and break on good cases.